core/src/avm2/optimize.rs

```diff
@@ -411,1179 +566,1466 @@ pub fn optimize<'gc>(
-    let mut state_map: HashMap<i32, Locals<'gc>> = HashMap::new();
+    // Map of op index -> local types + stack heights
+    let mut state_map: HashMap<i32, (Locals<'gc>, usize, usize)> = HashMap::new();
+    let mut expected_stack_heights: HashMap<i32, usize> = HashMap::new();
+    let mut expected_scope_stack_heights: HashMap<i32, usize> = HashMap::new();
 
-    let mut stack = Stack::new();
-    let mut scope_stack = Stack::new();
+    let mut stack = Stack::new(method_body.max_stack as usize);
+    let mut scope_stack =
+        ScopeStack::new((method_body.max_scope_depth - method_body.init_scope_depth) as usize);
     let mut local_types = initial_local_types.clone();
-    let mut last_op_was_block_terminating = false;
-
-    for (i, op) in code.iter_mut().enumerate() {
-        if let Some(jump_sources) = jump_targets.get(&(i as i32)) {
-            // Avoid handling multiple sources for now
-            if jump_sources.len() == 1 {
-                if let JumpSource::JumpFrom(source_i) = jump_sources[0] {
-                    // We can merge the locals easily, now
-                    if let Some(source_local_types) = state_map.get(&source_i) {
-                        let mut merged_types = initial_local_types.clone();
-                        assert_eq!(source_local_types.len(), local_types.len());
-
-                        if last_op_was_block_terminating {
-                            // If the last op was a block-terminating op, the
-                            // only possible way this is reachable is from
-                            // the jump. Just set the types to the types
-                            // at the jump.
-                            merged_types = source_local_types.clone();
-                        } else {
-                            for (i, target_local) in local_types.0.iter().enumerate() {
-                                let source_local = source_local_types.at(i);
+    let mut last_block_op_was_block_terminating = false;
+    let mut current_op_reached_linearly;
 
-                                merged_types.set(i, source_local.merged_with(*target_local));
+    let mut worklist = vec![0];
+
+    let mut seen_targets = HashSet::new();
+
+    while let Some(mut i) = worklist.pop() {
+        current_op_reached_linearly = false;
+        loop {
+            let last_op_was_block_terminating = if i != 0 {
+                code.get(i as usize - 1)
+                    .map_or(false, |o| o.is_block_terminating())
+            } else {
+                false
+            };
+
+            let op = &mut code[i as usize];
+
+            for exception in method_exceptions.iter() {
+                // If this op is in the to..from and it can throw an error,
+                // add the exception's target to the worklist.
+                if exception.from_offset as i32 <= i
+                    && i < exception.to_offset as i32
+                    && op.can_throw_error()
```

Comment:
you can check `op.can_throw_error()` before iterating exceptions

Reply:
The verifier has the same issue, I'll fix it in both places